cmd/tailscale/cli/web.go

```diff
@@ -228,33 +228,45 @@ func qnapAuthn(r *http.Request) (string, *qnapAuthResponse, error) {
 	return "", nil, fmt.Errorf("not authenticated by any mechanism")
 }
 
-func qnapAuthnQtoken(r *http.Request, user, token string) (string, *qnapAuthResponse, error) {
-	query := url.Values{
-		"qtoken": []string{token},
-		"user":   []string{user},
+// qnapAuthnURL returns the auth URL to use by inferring where the UI is
+// running based on the request URL. This is necessary because QNAP has so
+// many options, see https://github.com/tailscale/tailscale/issues/7108
+// and https://github.com/tailscale/tailscale/issues/6903
+func qnapAuthnURL(requestUrl string, query url.Values) string {
+	in, err := url.Parse(requestUrl)
+	scheme := in.Scheme
+	host := in.Host
```

Comment:
this will panic if `err != nil`

Reply:
I never found a condition where url.Parse returns an actual error, it instead returns err=nil and a struct with no fields populated. That is why the conditional includes `scheme == ""`, because that is what it returns when it cannot parse anything out of the URL.

Nonetheless it is a good point that tailscaled would crash if we ever did find a case where url.Parse returns an err. I think the updated code works: if err != nil the conditional will be short-circuited and in.Scheme won't be evaluated.